Functions/ImageProcessing/xASL_im_ASLSubtractionAveraging.m

```diff
@@ -0,0 +1,400 @@
+		nVolumes = size(ASL4Dnii.dat, 4);
+	else
+		nDims = length(size(ASL4D));
+		nVolumes = size(ASL4D, 4);
+	end
+	    
+    if nDims>4
+        fprintf('\n\nIn BIDS ASL NIfTIs should have [X Y Z n/PLD/TE/LD/etc] as 4 dimensions\n\n');
+        error('ASL4D.nii contains more than 4 dimensions');
+        
+    elseif nVolumes>1 && mod(nVolumes, 2) ~= 0 && ~x.modules.asl.bContainsSubtracted
+        warning('Odd (i.e., not even) number of control-label pairs, skipping');
+        % mod(nVolumes, 2) ~= 0 here means that nVolumes is not divisible by 2
+        % unsubtracted control-label pairs should result in an even number of volumes
+        return;
+    end
+
+
+    %% 2A. Apply motion correction if needed
+    % At the beginning of the ASL module we estimate motion, and optionally remove outliers/spikes
+    % This can already be applied, but in some cases this is not yet applied and by just loading ASL4D.nii
+    % this is then forgotten. So here, we check if a .mat (e.g., ASL4D.mat) sidecar exists, and then run
+    % resampling to apply motion correction. This creates a temporary r-prefixed file (e.g., rASL4D.nii)
+    % which is deleted right after we reload its NIfTI as image volumes.
+	if ~isempty(path_ASL4D)
+		[Fpath, Ffile] = xASL_fileparts(path_ASL4D);
+		sideCarMat = fullfile(Fpath, [Ffile '.mat']);
+
+		if exist(sideCarMat, 'file') && nVolumes>1
+			% we check if the mat-sidecar exists, with the motion estimation parameters
+			% If this file exists, we assume that the original ASL4D input file has not been
+			% interpolated yet (and the motion estimation has not been applied yet)
+
+			% Apply motion correction & resample
+
+			for iS=1:nVolumes
+				matlabbatch{1}.spm.spatial.realign.write.data{iS,1} = [path_ASL4D ',' num2str(iS)];
+			end
+
+			matlabbatch{1}.spm.spatial.realign.write.roptions.which     = [2 0];
+			matlabbatch{1}.spm.spatial.realign.write.roptions.interp    = 4;
+			matlabbatch{1}.spm.spatial.realign.write.roptions.wrap      = [0 0 0];
+			matlabbatch{1}.spm.spatial.realign.write.roptions.mask      = 1;
+			matlabbatch{1}.spm.spatial.realign.write.roptions.prefix    = 'r';
+
+			spm_jobman('run',matlabbatch); % this applies the motion correction in native space
+
+			path_rASL4D = fullfile(Fpath, ['r' Ffile '.nii']);
+
+			% RELOAD ASL4D.nii
+			% Note that we only reload ASL4D.nii here if we have motion corrected timeseries.
+			% If path_ASL4D was not a path, or if sideCarMat didn't exist (motion estimation was not performed)
+			% or if nVolumes==1 (we did not have multiple ASL volumes) there is no reason to resample and we keep using
+			% the already above loaded ASL4D.nii
+			ASL4D = xASL_io_Nifti2Im(path_rASL4D);
+			xASL_delete(path_rASL4D);
+		else
+			ASL4D = xASL_io_Nifti2Im(path_ASL4D);
+		end
+	end
+
+
+    %% 2B. Apply time decoding if needed
+    if x.modules.asl.bTimeEncoded
+        %% #997 Check that the motion correction of Hadamard is correctly performed here,
```

Comment:
This comment can be removed, right? As motion correction is done and resampled above.

Reply:
Yes, if you keep on using ASL4D as matrix. If you use a path, we have to change it potentially to path_rASL4D. Happy for you to fix/resolve.

Reply:
Yes - we use the previously loaded image, not the path. I will remove the comment and resolve myself.

Reply:
Done in 3b440c54